addons/web/static/src/js/core/misc.js

```diff
@@ -190,7 +190,8 @@ function displayNotification(parent, action) {
         title: _.escape(title),
         message: sprintf(_.escape(message), ...links),
         type,
-        sticky
+        sticky,
+        isMessageHtml: true, // dynamic parts of the message are escaped above
```

Comment:
If I understand correctly `isMessageHtml` only affects the message, the `title` is now always escaped. This means existing explicit title escapings (as with this call) must be removed or we'll have double-escaped titles.

Reply:
I'm still considering options here. From what I've seen so far, the rare case were we use html is to format a list, so I might as well allow a list as param instead.

But yes either way I would have to go over all existing occurrences doing an escape and removing the escape.

Reply:
> Anyway as per my first comment, I might remove the possibility to use html all together if all it is ever used for is to pass a list.

That's not the case e.g. some of the notifications bounced from the server (via the `display_notification` client action) contain links for instance. I know that @JKE-be also mentioned client-side notifications with various formatting, though I don't remember where so I don't have a link to provide.

Reply:
Fortunately the links are passed as a list from `display_notification` and not as html. So at the end it's just a list again, though indeed this would require a different list format to add the links compared to the more generic list.

Reply:
> Fortunately the links are passed as a list from `display_notification` and not as html. So at the end it's just a list again, though indeed this would require a different list format to add the links compared to the more generic list.

It's more complicated than that. The links are passed as a special list of items so that they can be properly escaped and merged in the message independently. The message is not actually a list of anything, it's text with links embedded.

Reply:
here an example that contains html in the notification

https://github.com/odoo/odoo/blob/14.0/addons/calendar/static/src/xml/notification_calendar.xml#L12

Not sure how you plan to make it easy to send a notification with link, button, ... with a list :/:
"You need to <b>configure</b> your api key here: <a>click here</a> or use another map provider.

You can choose <a>here</a> you default map provider"

Reply:
It doesn't change the point that it's possible to forward the links to the notification itself and make it part of its options. It's just moving the logic one level down, so that any html handling is done in the template with t-att/t-esc and nowhere else.

If there is a title display it, if there is a message display it, if there are links display them after the message, if there are list items display them below the rest with an actual list tag. This can be expanded to other needs too.

Advantage: no more html issues, consistent look for all notifications, easier code to maintain
Disadvantage: no easy customization (but do we really need it?)

I would also like to note we have an option to convert these notifications to native OS notifications, which don't support html anyway. Accepting only raw data and formatting them according to the destination would make handling this much easier.


That being said, thanks for the example in `calendar`. If we really want to support such cases, then we would definitely need to support a formatting language, maybe markdown as suggested by @mart-e. Or just keep the concept of `isMessageHtml` introduced in this PR, in current state html is still possible. As long as it is not html by default it's already much better.

Reply:
imo it is not because OS cannot show a bold word in a notification that it cannot be used to make it more readable in webclient. And it is frequent that web_editor show notification with link to backend setting to configure something. 
So yes, we should probably keep the html feature, even if opt-out by default, at least keep to support it when needed.

Reply:
Just an idea but another alternative to allow HTML in a safer way would be to accept a template name + parameters (or OWL equivalent component/props) to ensure the injected HTML comes from a template and not from manually built HTML.

Reply:
This is also very similar to the concept of slot in OWL https://github.com/odoo/owl/blob/master/doc/reference/slots.md that is used for example in Dialog and Popover, where Notification in that case would just be a "container" positioned appropriately and would receive its content from its parent template.

Reply:
> Just an idea but another alternative to allow HTML in a safer way would be to accept a template name + parameters (or OWL equivalent component/props) to ensure the injected HTML comes from a template and not from manually built HTML.

If the API accepts a literal template (not a template name but the actual contents of the template), it doesn't fix anything because you can trivially create an unsafe template. If the API does not accept a literal template, it can't do the job.

I was thinking of markdown for these situations, that allows for restricted customisations, there are libraries which allow enabling and disabling features as necessary, and it's trivial to disable the "HTML passthrough" feature. There are a few questions there (buttons, styling) but it does provide a good basis. And there are extensions available in existing markdown libraries e.g. https://www.npmjs.com/package/markdown-it-attrs

Reply:
I've removed the double escaped title here, as per the first suggestion. I haven't spotted another title that was already escaped, though sometimes it is a bit complex to figure out where the information is coming from...

Reply:
As for the rest of the discussion, I propose that we keep the proposition of current PR, which is making it safe by default, but still customizable with HTML by passing a special flag (that you can mark in the ci/security).

Transforming them markdown is out of scope for me, while I can see the added value, this should be another task.